### DISCUSSION TITLE
Development fragmentation

### DISCUSSION THREAD

**Discussion by trunglebka** [k2-fsa/sherpa#312] [General]
Currently we have 3 versions of sherpa: this repo, ncnn, onnx.
Do you think that this seperation may cause fragmentation?

(I don't have much expertise in C++ has not used ncnn, onnx versions yet but some time encountered this problem in general programing)

**Comment by csukuangfj** [comment]
sherpa, this repo, uses libtorch. It is targeted for x86 servers.

sherpa-ncnn uses ncnn.  It is mainly targed for embedded systems and mobile systems, e.g., raspberry pi, android, iOS.

sherpa-onnx uses onnxruntime. It can support both embedded systems and mobile systems as well as x86 servers

They are separated so that it simplifies the building system and also simplifies maintenance.

**Reply by trunglebka** [reply]
Thanks you. Just worry about maintenance burden but look like it is tied to ML framework too much to put together

**Comment by trunglebka** [comment]
@csukuangfj  Related to the OP - since there are many implementations, can you tell me which one has more priority? 
Currently I'm using this repo as starting point. I prefer solution that not being locked by Pytorch (due to installation), but I don't know if sherpa-onnx is keep pace with sherpa and that any performance gain/loss with sherpa-onnx.

**Reply by csukuangfj** [reply]
> since there are many implementations, can you tell me which one has more priority?

We would support all three: sherpa, sherpa-ncnn, and sherpa-onnx.

You can choose anyone you like.

**Reply by trunglebka** [reply]
I can see that from commit history, but which will get icefall's new recipes implement first?

**Reply by csukuangfj** [reply]
I would say sherpa.

Because exporting a model to sherpa is super straightforward, but it is not the case for sherpa-onnx and sherpa-ncnn.

**Reply by trunglebka** [reply]
Thanks, I asked this because the last time I checked I want to use streaming Zipformer but it had not yet implemented at that time and waiting is not happy (^3^)

**Reply by csukuangfj** [reply]
I think streaming zipformer (both v1 and v2) is already supported in sherpa. We are supporting sherpa-onnx right now.
(We have not exported the v2 model to sherpa-ncnn yet)
